otaclient/app/boot_control/_grub.py

```diff
@@ -578,201 +609,99 @@ def _ensure_ota_partition_symlinks(self):
+        except Exception as e:
+            _err_msg = f"failed to prepare standby dev: {e!r}"
+            raise BootControlPreUpdateFailed(_err_msg) from e
 
-        GrubController supports migrates system that doesn't boot via ota-partition
-        mechanism(possibly using different grub configuration, i.e., grub submenu enabled)
-        to use ota-partition.
+    def finalize_update_switch_boot(self):
+        """Finalize switch boot and use boot files from current booted slot."""
+        # NOTE: since we have not yet switched boot, the active/standby relationship is
+        #       reversed here corresponding to booted slot.
+        self._prepare_kernel_initrd_links(self.standby_ota_partition_folder)
+        self._ensure_ota_partition_symlinks(active_slot=self.standby_slot)
+        self._ensure_standby_slot_boot_files_symlinks(standby_slot=self.active_slot)
 
-        NOTE:
-        1. only update the ota-partition.<active_slot>/grub.cfg!
-        2. standby slot is not considered here!
-        3. expected previously booted kernel/initrd to be located at /boot
-        """
-        # check the current booted kernel,
-        # if it is not vmlinuz-ota, copy that kernel to active ota_partition folder
-        cur_kernel, cur_initrd = self._get_current_booted_kernel_and_initrd()
-        if cur_kernel != GrubHelper.KERNEL_OTA or cur_initrd != GrubHelper.INITRD_OTA:
-            logger.info(
-                "system doesn't use ota-partition mechanism to boot, "
-                "initializing ota-partition file..."
-            )
-            # NOTE: just copy but not cleanup the existed kernel/initrd files
-            shutil.copy(
-                self.boot_dir / cur_kernel,
-                self.active_ota_partition_folder,
-                follow_symlinks=True,
-            )
-            shutil.copy(
-                self.boot_dir / cur_initrd,
-                self.active_ota_partition_folder,
-                follow_symlinks=True,
-            )
-            self.reprepare_active_ota_partition_file(abort_on_standby_missed=False)
+        self._grub_update_on_booted_slot(abort_on_standby_missed=True)
 
-        logger.info("ota-partition file initialized")
+        # switch ota-partition symlink to current booted slot
+        self._ensure_ota_partition_symlinks(active_slot=self.active_slot)
+        self._ensure_standby_slot_boot_files_symlinks(standby_slot=self.standby_slot)
+        return True
 
     def grub_reboot_to_standby(self):
-        self.reprepare_standby_ota_partition_file()
+        """Temporarily boot to standby slot after OTA applied to standby slot."""
+        self._prepare_kernel_initrd_links(self.standby_ota_partition_folder)
+        self._ensure_standby_slot_boot_files_symlinks(standby_slot=self.standby_slot)
+
+        self._prepare_kernel_initrd_links(self.active_ota_partition_folder)
+        self._ensure_ota_partition_symlinks(active_slot=self.active_slot)
```

Comment:
I'd like to know why these two lines are needed?

Reply:
These two methods are to ensure the required symlinks are presented **at the time just before grub-reboot is called**. 

If required symlinks are presented, these two methods won't do others things, otherwise they will re-generate the missing symlinks.

Reply:
FYI, the `re_symlink_atomic` method used by these two methods will check the symlink before actually creating symlink again. So if the symlink is presented and already points to correct target, this method will do nothing.

Reply:
Additional comments for this method are added at [ebb0c4c](https://github.com/tier4/ota-client/pull/257/commits/ebb0c4c93bc1a1b49eab3bdcd893b621611e503a).